Schedule Local Notification for Product Image Uploads if the app is sent to background

pmusolino opened the pull request:
## Description
This PR introduces a new feature that schedules a local notification when the app is sent to the background with active product image uploads. This ensures that users are informed about ongoing uploads and potential issues due to background processing.

## Testing information
- Upload many new images into a product and send the app into the background. You should be able, on a device, to see the local notification.
- Upload many new images into a product, save, and go back or to a different screen. Then send the app into the background. You should be able, on a device, to see the local notification.

## Screenshots
<img src="https://github.com/user-attachments/assets/59c7f24b-75bc-460d-824e-f44e817b6c76" width=300 />


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

pmusolino committed:
feat: new LocalNotification type

pmusolino committed:
Feat: Schedule a local notification if the app has some image uploads and the app is sent into the background

pmusolino committed:
update: RELEASE-NOTES 21.8

pmusolino committed:
fix: private visibility of `scheduleUploadInProgressNotification` method

pmusolino committed:
update: added PR link

pmusolino committed:
fix: lint issues

itsmeichigo commented on WooCommerce/Classes/Notifications/LocalNotification.swift:
```diff
@@ -128,5 +134,18 @@ extension LocalNotification {
+        enum ProductImageBackgroundUpload {
+            static let title = NSLocalizedString(
+                "localNotification.ProductImageUploader.title",
+                value: "Product Image Upload in Progress",
```
For a more concise message, how about "Uploading in Progress"?

pmusolino replied:
I think if we change it to "uploading in progress," there will be a subject missing, like uploading of what? So, I'm changing it to "Image Upload in Progress." Please, let me know what you think.

itsmeichigo replied:
I'm not a native English speaker so I'm unsure as well 😅 Let's keep this for now then.

itsmeichigo commented on WooCommerce/Classes/Notifications/LocalNotification.swift:
```diff
@@ -128,5 +134,18 @@ extension LocalNotification {
+            )
+            static let body = NSLocalizedString(
+                "localNotification.ProductImageUploader.message",
+                value: "Your product images are still uploading, but upload speed may be reduced" +
+                "and errors could occur. For best results, please keep the app open until uploads finish.",
```
A space is missing between "reduced" and "and".

I also asked AI for a more concise message, and this is what's suggested:

> Your product images are still uploading in the background. Upload speed may be slower, and errors could occur. For best results, please keep the app open until uploads are complete.

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -243,11 +254,35 @@ final class ProductImageUploader: ProductImageUploaderProtocol {
+    }
+
+    @objc private func appDidEnterBackground() {
+        activeUploads
+            .first()
+            .sink { [weak self] activeKeys in
+                guard let self = self else { return }
+                if !activeKeys.isEmpty {
+                    self.scheduleUploadInProgressNotification()
+                }
+            }
+            .store(in: &backgroundSubscriptions)
```
Any reason why we need to observe the first value of `activeUploads` instead of triggering `scheduleUploadInProgressNotification` immediately?

itsmeichigo reviewed (commented):
Thank you Paolo for this improvement! I added suggestions for clearer messages and left a question below.

pmusolino committed:
Merge branch 'trunk' into feat/local-notification-if-image-upload-is-in-progress-and-app-is-sent-to-background

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/238">21.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino committed:
update: localized strings

pmusolino committed:
feat: remove combine subscription and replace with a direct check on `activeUploadsPublisher`

pmusolino commented:
This is ready for another check

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -112,6 +113,7 @@ final class ProductImageUploader: ProductImageUploaderProtocol {
     private let errorsSubject: PassthroughSubject<ProductImageUploadErrorInfo, Never> = .init()
     private var statusUpdatesExcludedProductKeys: Set<Key> = []
     private var statusUpdatesSubscriptions: Set<AnyCancellable> = []
+    private var backgroundSubscriptions: Set<AnyCancellable> = []
```
This is unused now, so it should be removed.

itsmeichigo commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -243,12 +254,30 @@ final class ProductImageUploader: ProductImageUploaderProtocol {
+    }
+
+    @objc private func appDidEnterBackground() {
+        if !activeUploadsPublisher.isEmpty {
+            self.scheduleUploadInProgressNotification()
+        }
```
Nit: since you're checking `activeUploadsPublisher` in `scheduleUploadInProgressNotification`, this extra check seems redundant. I'd suggest removing the check and renaming the method to `scheduleUploadInProgressNotificationIfNeeded`

itsmeichigo reviewed (approved):
Thank you for the updates, I left some comments but approving the PR now. Please feel free to merge after addressing the issues.

pmusolino committed:
- Remove unused `backgroundSubscriptions` property in `ProductImageUploader.swift`
- Rename `scheduleUploadInProgressNotification` to `scheduleUploadInProgressNotificationIfNeeded` for clarity

pmusolino commented:
Thanks, Huong! I handled your suggestions, and I'm going to merge them now.

pmusolino committed:
Merge branch 'trunk' into feat/local-notification-if-image-upload-is-in-progress-and-app-is-sent-to-background
